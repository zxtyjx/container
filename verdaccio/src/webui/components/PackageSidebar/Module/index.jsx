import React from 'react';
import PropTypes from 'prop-types';

import classes from './style.scss';

export default function Module({title, description, children, className}) {
  return (
    <div className={`${classes.module} ${className}`}>
      <h2 className={classes.moduleTitle}>
        {title}
        {description && <span>{description}</span>}
      </h2>
      <div>
        {children}
      </div>
    </div>
  );
}

Module.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  children: PropTypes.any.isRequired,
  className: PropTypes.string
};
