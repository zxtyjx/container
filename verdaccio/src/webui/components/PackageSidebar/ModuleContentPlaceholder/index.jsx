import React from 'react';
import PropTypes from 'prop-types';

import classes from './style.scss';

export default function ModuleContentPlaceholder({text}) {
  return <p className={classes.emptyPlaceholder}>{text}</p>;
}
ModuleContentPlaceholder.propTypes = {
  text: PropTypes.string.isRequired
};
