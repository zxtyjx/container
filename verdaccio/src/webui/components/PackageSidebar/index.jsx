import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import LastSync from './modules/LastSync';
import Maintainers from './modules/Maintainers';
import Dependencies from './modules/Dependencies';
import Infos from './modules/Infos';

import {
  formatLicense,
  formatRepository,
  getLastUpdatedPackageTime,
  getRecentReleases
} from '../../utils/package';
import API from '../../utils/api';

export default class PackageSidebar extends React.Component {
  state = {};

  static propTypes = {
    packageName: PropTypes.string.isRequired
  };

  constructor(props) {
    super(props);
    this.loadPackageData = this.loadPackageData.bind(this);
  }

  async componentDidMount() {
    await this.loadPackageData(this.props.packageName);
  }

  async UNSAFE_componentWillReceiveProps(newProps) {
    if (newProps.packageName !== this.props.packageName) {
      await this.loadPackageData(newProps.packageName);
    }
  }

  async loadPackageData(packageName) {
    let packageMeta;

    try {
      packageMeta = await API.request(`sidebar/${packageName}`, 'GET');
    } catch (err) {
      this.setState({
        failed: true
      });
    }

    this.setState({
      packageMeta
    });
  }

  render() {
    let {packageMeta} = this.state;

    if (packageMeta) {
      const {time, _uplinks} = packageMeta;

      // Infos component
      const license = formatLicense(get(packageMeta, 'latest.license', null));
      const repository = formatRepository(
        get(packageMeta, 'latest.repository', null)
      );
      const homepage = get(packageMeta, 'latest.homepage', null);

      // Lastsync component
      const recentReleases = getRecentReleases(time);
      const lastUpdated = getLastUpdatedPackageTime(_uplinks);

      // Dependencies component
      const dependencies = get(packageMeta, 'latest.dependencies', {});

      // Maintainers component
      return (
        <aside className="sidebar-info">
          {time && (
            <LastSync
              recentReleases={recentReleases}
              lastUpdated={lastUpdated}
            />
          )}
          <Infos
            homepage={homepage}
            repository={repository}
            license={license}
          />
          {/* TODO: Refacor later, when we decide to show only maintainers/authors */}
          <Maintainers packageMeta={packageMeta} />
          <Dependencies dependencies={dependencies} />
          {/* Package management module? Help us implement it! */}
        </aside>
      );
    }
    return (
      <aside className="sidebar-loading">Loading package information...</aside>
    );
  }
}
