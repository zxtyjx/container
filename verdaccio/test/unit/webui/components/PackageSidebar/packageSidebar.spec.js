/**
 * Package Sidebar module
 */

import React from 'react';
import { mount } from 'enzyme';
import PackageSidebar from '../../../../../src/webui/components/PackageSidebar/index';
import { packageMeta } from '../store/packageMeta';

jest.mock('../../../../../src/webui/utils/api', () => ({
  request: require('../__mocks__/api').default.request,
}));

console.error = jest.fn();

describe('<PackageSidebar /> component', () => {
  it('should throw error for the required props', () => {
    const wrapper = mount(<PackageSidebar />);
    const { loadPackageData } = wrapper.instance();
    expect(console.error).toBeCalled();
    loadPackageData().catch(response => {
      expect(response).toBeUndefined();
      expect(wrapper.state()).toEqual({ failed: true });
    });
  });

  it('should load the packageMeta', () => {
    const wrapper = mount(<PackageSidebar packageName={'verdaccio'} />);
    const { loadPackageData } = wrapper.instance();
    loadPackageData('verdaccio').then(response => {
      expect(wrapper.state('packageMeta')).toEqual(packageMeta);
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
