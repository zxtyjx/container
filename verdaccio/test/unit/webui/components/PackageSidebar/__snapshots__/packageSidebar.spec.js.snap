// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackageSidebar /> component should load the packageMeta 1`] = `"<aside class=\\"sidebar-loading\\">Loading package information...</aside>"`;
